### PR TITLE
Initial support of parametric hardware models

### PR DESCRIPTION
**Pull request by diegoferigo**
Currently, JaxSim builds three models from a model description (URDF, SDF):

- [`jaxsim.parsers.descriptions.ModelDescription`](https://github.com/ami-iit/jaxsim/blob/4fd2032c74de495cd59e1f6999e1f7c3c9ec2ff8/src/jaxsim/parsers/descriptions/model.py#L14)
- [`jaxsim.physics.model.PhysicsModel`](https://github.com/ami-iit/jaxsim/blob/4fd2032c74de495cd59e1f6999e1f7c3c9ec2ff8/src/jaxsim/physics/model/physics_model.py#L23)
- [`jaxsim.api.model.JaxSimModel`](https://github.com/ami-iit/jaxsim/blob/4fd2032c74de495cd59e1f6999e1f7c3c9ec2ff8/src/jaxsim/api/model.py#L28) (functional replacement of the former [`jaxsim.high_level.model.Model`](https://github.com/ami-iit/jaxsim/blob/4fd2032c74de495cd59e1f6999e1f7c3c9ec2ff8/src/jaxsim/high_level/model.py#L97))

One of the objective of the new functional APIs (#108) is to expose a set of hardware parameters so that downstream projects can compute their gradient by exploiting automatic differentiation.

This PR introduces the initial support towards this goal. Particularly:

- Introduces a new `KinDynParameters` class, stored inside `JaxSimModel`, that will soon replace `PhysicsModel`. Its aim is to act as a simple container of joint, link, and contact parameters, with useful setters for updating them without re-creating from scratch a new `JaxSimModel` object.
- Starts the transition from `PhysicsModel` to `JaxSimModel` (will be finalized in [#112](https://github.com/ami-iit/jaxsim/pull/112)).

----
📚 Documentation preview 📚: https://jaxsim--101.org.readthedocs.build//101/

### CODE REVIEW

**Comment by S-Dafarra** [comment]
cc @CarlottaSartore

File: src/jaxsim/api/ode.py [hunk 1]
```diff
@@ -198,7 +202,7 @@ def system_velocity_dynamics(
 
 @jax.jit
 def system_position_dynamics(
-    model: Model.JaxSimModel, data: Data.JaxSimModelData
+    model: js.model.JaxSimModel, data: js.data.JaxSimModelData
```
Comment (flferretti) [inline src/jaxsim/api/ode.py:205]: `model` is not used
```suggestion
    data: js.data.JaxSimModelData
```

File: src/jaxsim/api/ode.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 import jax.numpy as jnp
 import jaxlie
 
+import jaxsim.api as js
 import jaxsim.physics.algos.soft_contacts
 import jaxsim.typing as jtp
 from jaxsim import VelRepr, integrators
```
Comment (flferretti) [inline src/jaxsim/api/ode.py:10]: `jaxlie` and `integrators` are not used
```suggestion

import jaxsim.api as js
import jaxsim.physics.algos.soft_contacts
import jaxsim.typing as jtp
from jaxsim import VelRepr
```

File: src/jaxsim/physics/algos/soft_contacts.py [hunk 3]
```diff
@@ -247,11 +247,13 @@ def process_point_kinematics(
...
 
     # Process all the collidable points in parallel
-    W_p_Ci, CW_v_WC = jax.vmap(process_point_kinematics)(
-        model.gc.point.T, np.array(model.gc.body, dtype=int)
-    )
+    W_p_Ci, CW_vl_WC = jax.vmap(
+        lambda Li_p_C, parent_body: process_point_kinematics(
+            Li_p_C=Li_p_C, parent_body=parent_body
+        )
+    )(model.gc.point, jnp.array(model.gc.body))
```
Comment (flferretti) [inline src/jaxsim/physics/algos/soft_contacts.py:254]: Is the lambda necessary?
```suggestion
    W_p_Ci, CW_vl_WC = jax.vmap(process_point_kinematics)(Li_p_C, parent_body)(
        model.gc.point, jnp.array(model.gc.body)
    )
```

**Review by flferretti — APPROVED** [review]
Thanks Diego, LGTM! I just wrote some minor comments

File: src/jaxsim/api/kin_dyn_parameters.py [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+            mass=jnp.array(m).squeeze().astype(float),
+            center_of_mass=jnp.atleast_1d(jnp.array(L_p_CoM).squeeze()).astype(float),
+            inertia_elements=jnp.atleast_1d(I[jnp.triu_indices(3)].squeeze()).astype(
+                float
+            ),
+        )
+
+    @staticmethod
+    def build_from_inertial_parameters(
+        m: mass, I: jtp.MatrixLike, c: jtp.VectorLike
```
[the remark this answers is not in the document]
Comment (diegoferigo) [reply src/jaxsim/api/kin_dyn_parameters.py:460]:
```suggestion
        m: jtp.FloatLike, I: jtp.MatrixLike, c: jtp.VectorLike
```

File: src/jaxsim/api/kin_dyn_parameters.py [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+    ) -> LinkParameters:
+
+        return LinkParameters(
+            mass=jnp.array(m).squeeze().astype(float),
+            I=jnp.atleast_1d(I[jnp.triu_indices(3)].squeeze()).astype(float),
+            com=jnp.atleast_1d(c.squeeze()).astype(float),
```
[the remark this answers is not in the document]
Comment (diegoferigo) [reply src/jaxsim/api/kin_dyn_parameters.py:466]: Good catch! I still don't have tests against these builders. Will be included as soon as I try running AD on hardware parameters.

File: src/jaxsim/api/ode.py (hunk 1 above)
Comment (diegoferigo) [reply src/jaxsim/api/ode.py:205]: Waiting the semi-implicit integrators, this is intended. I'll leave it as it is in the `functional` branch, let's discuss it again before merging into `main`.

File: src/jaxsim/physics/algos/soft_contacts.py (hunk 3 above)
Comment (diegoferigo) [reply src/jaxsim/physics/algos/soft_contacts.py:254]: It is not necessary but PyCharm raises a warning (that I could ignore). I remove the lambda.

File: src/jaxsim/api/kin_dyn_parameters.py [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+            )
+
+            return κb
+
+        support_body_array_bool = jax.vmap(κb)(
+            jnp.arange(start=0, stop=len(ordered_links))
+        )
+
+        return KynDynParameters(
+            link_names=tuple([l.name for l in ordered_links]),
```
[the remark this answers is not in the document]
Comment (diegoferigo) [reply src/jaxsim/api/kin_dyn_parameters.py:153]:
```suggestion
        m: jtp.FloatLike, I: jtp.MatrixLike, c: jtp.VectorLike
```

File: src/jaxsim/physics/algos/soft_contacts.py (hunk 3 above)
Comment (flferretti) [reply src/jaxsim/physics/algos/soft_contacts.py]: Alright! Just for curiosity, which warning is raised?

Comment (diegoferigo) [reply src/jaxsim/physics/algos/soft_contacts.py]: It's just PyCharm that fails sometimes to understand that the arguments of a generic `jax.vmap(my_fn)` are those of `my_fn`. Defining a lambda helps the IDE in this case. Not a big deal.

Comment (diegoferigo) [reply src/jaxsim/physics/algos/soft_contacts.py]: For the records, the pattern using a lambda is quite delicate if it would be jax.jit instead of jax.vmap. In fact, a lambda created on-the-fly has a different `id` that re-triggers jit recompilations.
